### PR TITLE
[ENH] Add metadata fields for DeIdentificationMethod/CodeSequence for MRI and PET

### PR DESCRIPTION
**Pull request by CGSchwarzMayo**
Add tags for DeIdentificationMethod and DeIdentificationMethodCodeSeuence, corresponding to the same-named DICOM tags. These record info on de-identification, including metadata and image content (e.g. de-facing or removing burned-in text)

### CODE REVIEW

**Comment by CGSchwarzMayo** [comment]
This addresses issue #1709

File: src/schema/objects/metadata.yaml
Comment (effigies) [inline src/schema/objects/metadata.yaml:593]: Fleshing out these values with further descriptions, so we can make a table describing these fields.

```suggestion
      - CodeValue
      - CodeMeaning
      - CodingSchemeDesignator
      - CodingSchemeVersion
    properties:
      CodeValue:
        name: CodeValue
        type: string
        description: |
          An identifier that is unambiguous within the Coding Scheme
          denoted by Coding Scheme Designator and Coding Scheme Version.
          Corresponds to DICOM Tag 0008, 0100 `Code Value`.
      CodeMeaning:
        name: CodeMeaning
        type: string
        description: |
          Text that has meaning to a human and conveys the meaning of the term
          Corresponds to DICOM Tag 0008, 0104 `Code Meaning`.
      CodingSchemeDesignator:
        name: CodingSchemeDesignator
        type: string
        description: |
          The identifier of the coding scheme in which the coded entry is defined.
          Corresponds to DICOM Tag 0008, 0102 `Coding Scheme Designator`.
      CodingSchemeVersion:
        name: CodingSchemeVersion
        type: string
        description: |
          An identifier of the version of the coding scheme if necessary to resolve ambiguity.
          Corresponds to DICOM Tag 0008, 0103 `Coding Scheme Version`.
```

**Review by effigies — COMMENTED** [review]
Okay, to validate these, we will need to add them to sidecars.

https://github.com/bids-standard/bids-specification/blob/master/src/schema/rules/sidecars/mri.yaml

I would suggest a pair of new tables, such as:

```yaml
DeidentificationMethod:
  selectors:
    - intersects([modality], ["mri", "pet"])
  fields:
    DeidentificationMethod: optional
    DeidentificationMethodCodeSequence:
      level: optional, but recommended if DeidentificationMethod is present

DeidentificationMethodRec:
  selectors:
    - intersects([modality], ["mri", "pet"])
    - sidecar.DeidentificationMethod
  fields:
    DeidentificationMethodCodeSequence: recommended
```

Then for rendering, these can then be added to the lists of tables for MRI data:

https://github.com/bids-standard/bids-specification/blob/d7b37931c3e3754904611b6e6f64b7da8932443c/src/modality-specific-files/magnetic-resonance-imaging-data.md?plain=1#L16-L176

e.g.

```markdown
### Deidentification information

<!-- This block generates a metadata table.
These tables are defined in
  src/schema/rules/sidecars
The definitions of the fields specified in these tables may be found in
  src/schema/objects/metadata.yaml
A guide for using macros can be found at
 https://github.com/bids-standard/bids-specification/blob/master/macros_doc.md
-->
{{ MACROS___make_sidecar_table("mri.DeidentificationMethod") }}

Each object in the `DeidentificationMethodCodeSequence` array includes the following RECOMMENDED keys:

<!-- This block generates a table describing subfields within a metadata field.
The definitions of these fields can be found in
  src/schema/objects/metadata.yaml
and a guide for using macros can be found at
 https://github.com/bids-standard/bids-specification/blob/master/macros_doc.md
-->
{{ MACROS___make_subobject_table("metadata.DeidentificationMethodCodeSequence.items") }}
```

And we should do the same for PET.

**Comment by CGSchwarzMayo** [comment]
I apologize for missing some pieces, and I think I understand, now. Thank you for your careful review and for helping me through this process. I will execute your suggestions and amend the PR with another commit.

**Comment by effigies** [comment]
No need to apologize! I appreciate the effort you're putting into this.

**Comment by CGSchwarzMayo** [comment]
Chris,

I committed your suggested diffs as-is. For the tables and sidecars, I deviated a bit in calling both tags optional. My read of DICOM is that if PatientIdentityRemoved is TRUE (which is a tag we don't have in BIDS, and I don't think BIDS really needs it), then at least one of either DeidentificationMethod or DeidentificationCodeSequence is required. Without PatientIdentityRemoved, I would say neither is required, and both are optional. It's not that DeidentificationMethod is recommended if DeidentificationCodeSequence is present, but rather they are of equal status and people can use either. DeidentificationMethod is a single simple string, but there's a greater chance that some downstream software would replace rather than amend the string. DeidentificationCodeSequence better accounts for there being a series of de-identification steps and other software would more likely amend than overwrite it, but it's also a sequence tag and those are unpleasantly messy. 

While I was copying from the MRI to the PET table, I noticed the "defacemask" block in the MRI but not PET. I copied this to the PET, since de-facing is also routinely applied to PET now. 

I looked for CT-related files, but it appears that CT is not a part of BIDS currently? If CT is ever added, my recommendation is that these elements all be added to CT to match MRI and PET.

Thanks for your help and careful review!

File: src/modality-specific-files/positron-emission-tomography.md
Comment (effigies) [inline src/modality-specific-files/positron-emission-tomography.md:482]: Quick fixup to resolve the rendering failure.

```suggestion
      ])
}}
```

**Comment by effigies** [comment]
> For the tables and sidecars, I deviated a bit in calling both tags optional. My read of DICOM is that if PatientIdentityRemoved is TRUE (which is a tag we don't have in BIDS, and I don't think BIDS really needs it), then at least one of either DeidentificationMethod or DeidentificationCodeSequence is required. Without PatientIdentityRemoved, I would say neither is required, and both are optional.

Sounds good.

> While I was copying from the MRI to the PET table, I noticed the "defacemask" block in the MRI but not PET. I copied this to the PET, since de-facing is also routinely applied to PET now.

Could you split that out into a separate PR? I think it will need some additional work to apply properly to PET, and I'd rather not complicate this PR more than necessary.

> I looked for CT-related files, but it appears that CT is not a part of BIDS currently? If CT is ever added, my recommendation is that these elements all be added to CT to match MRI and PET.

Correct, and agreed!

File: src/modality-specific-files/magnetic-resonance-imaging-data.md
Comment (effigies) [inline src/modality-specific-files/magnetic-resonance-imaging-data.md:384]: I think we should add this after `### Tissue description` (L165-L176), since this is metadata that could in principle apply to any image, not just anatomical images. While you're at it, would you mind removing one of the hash marks (`#`) from "Tissue description", since it's incorrectly being interpreted as a subsection of "Echo-Planar Imaging and B0 mapping":

image

File: src/modality-specific-files/positron-emission-tomography.md
Comment (effigies) [inline src/modality-specific-files/positron-emission-tomography.md:460]: Likewise, let's move this under `#### Tissue description` above. And we will need four hash marks (`#### Deidentification information`) there.

File: src/schema/rules/sidecars/pet.yaml
Comment (effigies) [inline src/schema/rules/sidecars/pet.yaml:289]: We don't actually need this duplicated, since the `- intersects(...)` selector above ensures it will apply to MRI and PET.

```suggestion
```

**Review by effigies — COMMENTED** [review]
A few small suggestions. Also, in case you haven't seen it hiding in the review boxes (see image), here's the [rendered specification](https://bids-specification--1772.org.readthedocs.build/en/1772/), which will update a few minutes after you make pushes to this PR.

image

**Comment by CGSchwarzMayo** [comment]
Chris,
I attempted to make all the changes you suggested. The automated checks that failed appear to be related to parts of the code that I did not change in this branch, so unrelated to my changes. Thanks for your continuing reviews. 
Chris

**Review by effigies — APPROVED** [review]
I'm happy with this from a schema and rendering perspective, and I believe this should be relatively simple to implement for DICOM converters.

I would appreciate if at least a couple representatives of converters could weigh in:

* @yarikoptic 
* @marcelzwiers
* @dlevitas
* @neurolabusc

(I know you're all busy, please feel free to delegate.)

For those who would rather read spec than schema:

* [Magnetic Resonance Imaging - Deidentification information](https://bids-specification--1772.org.readthedocs.build/en/1772/modality-specific-files/magnetic-resonance-imaging-data.html#deidentification-information)
* [Positron Emission Tomography - Deidentification information](https://bids-specification--1772.org.readthedocs.build/en/1772/modality-specific-files/positron-emission-tomography.html#deidentification-information)

**Comment by CGSchwarzMayo** [comment]
Thanks Chris (Markiewicz)! 

I have a companion PR for dcm2niix at https://github.com/rordenlab/dcm2niix/pull/813, but Chris (Rorden) hasn't commented on it yet. I had initially asked him to add the tags to dcm2niix's .json, and he said to get them added to bids-specification first, so that's how arrived here. 

I don't have any involvement with the other converters. 

Thanks,
Chris (Schwarz)

File: src/schema/objects/metadata.yaml
Comment (yarikoptic) [inline src/schema/objects/metadata.yaml:610]: I think this is complicated enough warranting addition to https://github.com/bids-standard/bids-examples if not as an example here in the BIDS spec text.  Do we have some nice example DICOMs to check those fields out?

Comment (CGSchwarzMayo) [reply src/schema/objects/metadata.yaml:610]: I think if someone were creating these tags from scratch at the BIDS level, they would use DeidentificationMethod rather than DeidentificationMethodCodeSequence. The latter is more provided to allow direct translation by dcm->nii and dcm->bids software. 

I'm attaching some nice example DICOM, as requested. This is MRI of a spaghetti squash (thanks @captainnova), and I've further de-identified it, so there's no PHI. I included the tags as if I had de-faced it, but it is a squash, so there was no face to de-face. 

[dcm_fruit_deidentified.zip](https://github.com/bids-standard/bids-specification/files/15122386/dcm_fruit_deidentified.zip)


The relevant dcm tags look like this:
```
(0012, 0063) De-identification Method            LO: ['mri_reface 0.3.4', 'Per DICOM PS 3.15 AnnexE. Details in 0012,0064']
(0012, 0064)  De-identification Method Code Sequence   5 item(s) ---- 
   (0008, 0100) Code Value                          SH: '113102'
   (0008, 0102) Coding Scheme Designator            SH: 'DCM'
   (0008, 0103) Coding Scheme Version               SH: '01'
   (0008, 0104) Code Meaning                        LO: 'Clean Recognizable Visual Features Option'
   ---------
   (0008, 0100) Code Value                          SH: 'replace_recognizable'
   (0008, 0102) Coding Scheme Designator            SH: 'mri_reface'
   (0008, 0103) Coding Scheme Version               SH: '0.3.4'
   (0008, 0104) Code Meaning                        LO: 'Replace face, ears, and artifacts in air'
   ---------
   (0008, 0100) Code Value                          SH: '113100'
   (0008, 0102) Coding Scheme Designator            SH: 'DCM'
   (0008, 0104) Code Meaning                        LO: 'Basic Application Confidentiality Profile'
   ---------
   (0008, 0100) Code Value                          SH: '113107'
   (0008, 0102) Coding Scheme Designator            SH: 'DCM'
   (0008, 0104) Code Meaning                        LO: 'Retain Longitudinal Temporal Information Modified Dates Option'
   ---------
   (0008, 0100) Code Value                          SH: '113111'
   (0008, 0102) Coding Scheme Designator            SH: 'DCM'
   (0008, 0104) Code Meaning                        LO: 'Retain Safe Private Option'
   ---------
```


With https://github.com/rordenlab/dcm2niix/pull/813, the relevant portion of json output is:
```
        "DeidentificationMethod": "mri_reface 0.3.4\\Per DICOM PS 3.15 AnnexE. Details in 0012,0064",
        "DeidentificationMethodCodeSequence": [ 
          { 
                "CodeValue": "113102",
                "CodingSchemeDesignator": "DCM",
                "CodingSchemeVersion": "01",
                "CodeMeaning": "Clean Recognizable Visual Features Option"
          },
          { 
                "CodeValue": "replace_recognizable",
                "CodingSchemeDesignator": "mri_reface",
                "CodingSchemeVersion": "0.3.4",
                "CodeMeaning": "Replace face, ears, and artifacts in air"
          },
          { 
                "CodeValue": "113100",
                "CodingSchemeDesignator": "DCM",
                "CodeMeaning": "Basic Application Confidentiality Profile"
          },
          { 
                "CodeValue": "113107",
                "CodingSchemeDesignator": "DCM",
                "CodeMeaning": "Retain Longitudinal Temporal Information Modified Dates Option"
          },
          { 
                "CodeValue": "113111",
                "CodingSchemeDesignator": "DCM",
                "CodeMeaning": "Retain Safe Private Option"
          }
        ],
```

**Comment by effigies** [comment]
I just want to check in on the status here. It looks like there aren't objections, but are we waiting on a PR to bids-examples, or an agreement in principle RE https://github.com/rordenlab/dcm2niix/pull/813?

File: src/schema/objects/metadata.yaml
Comment (effigies) [inline src/schema/objects/metadata.yaml:570]: Following up on https://github.com/rordenlab/dcm2niix/pull/813#discussion_r1674566934.

I'm not sure if this practically shows up in DICOM as a list of strings, or if it is either a string or a list of strings.

If we want to generally make this an array of strings, it would be:

```suggestion
  type: array
  items:
    type: string
```

If it would be a string or an array of strings:

```suggestion
  anyOf:
    - type: string
    - type: array
      items:
        type: string
```

I tend to prefer the simpler array of strings, rather than allowing it to sometimes be a string, but I'm okay following what DICOM does, for consistency.

Comment (bendhouseart) [reply src/schema/objects/metadata.yaml:570]: Would be happy to approve this and get it merged once the type is updated per @effigies suggestion.

Comment (CGSchwarzMayo) [reply src/schema/objects/metadata.yaml:570]: I am at a conference this week. I can make the change next week when I return.
Thank you both for your review!

**Comment by bendhouseart** [comment]
This looks great as it would greatly speed up my ability to determine if PET images have been defaced or not. Will very happily add this to [PET deface](https://github.com/openneuropet/petdeface) as soon as it's merged.

**Review by bendhouseart — APPROVED** [review]
LGTM thank you for submitting this!

**Pull request merged**
